### PR TITLE
Group membership checks

### PR DESCRIPTION
**Pull request by dtebbs**
Safe subgroup membership check methods.  These methods are generic, relying on large scalar multiplications.  More optimal per-curve checks may be added in a future PR.

### CODE REVIEW

**Comment by AntoineRondelet** [comment]
Related to: https://github.com/clearmatics/libff/issues/16

**Comment by dtebbs** [comment]
It could be problematic to do these checks by default when deserializing.  At least with the naive versions here.
Deserialising anything sizable is unfeasible, so we may have to default to no checks and add `read_*_safe`  functions as per the original description.

**Comment by dtebbs** [comment]
Updated this PR to include only the addition of the (unoptimized) `is_in_safe_subgroup`. Performing these checks on load can be addressed as part of another PR.

File: libff/algebra/curves/alt_bn128/alt_bn128_g1.cpp [hunk 1]
```diff
@@ -396,6 +398,11 @@ bool alt_bn128_G1::is_well_formed() const
     }
 }
 
+bool alt_bn128_G1::is_in_safe_subgroup() const
+{
+    return true;
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libff/algebra/curves/alt_bn128/alt_bn128_g1.cpp:404]: Added a comment to explain.

File: libff/algebra/curves/tests/test_groups.cpp [hunk 2]
```diff
@@ -155,6 +155,78 @@ void test_output()
...
+        assert(g.is_in_safe_subgroup());
+    }
+}
+
+template<typename GroupT>
+void test_group_membership_invalid_g1(const typename GroupT::base_field &x)
+{
+    const typename  GroupT::base_field x_squared = x * x;
+    const typename  GroupT::base_field x_cubed = x_squared * x;
+    const typename  GroupT::base_field y_squared =
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libff/algebra/curves/tests/test_groups.cpp]: removed spaces

File: libff/algebra/curves/tests/test_groups.cpp [hunk 3]
```diff
@@ -155,6 +155,78 @@ void test_output()
...
+    test_group_membership_valid<G1<ppT>>();
+    test_group_membership_valid<G2<ppT>>();
+}
+
+template<>
+void test_check_membership<alt_bn128_pp>()
+{
+    test_group_membership_valid<alt_bn128_G1>();
+    test_group_membership_valid<alt_bn128_G2>();
+    test_group_membership_invalid_g2<alt_bn128_G2>(alt_bn128_Fq2::one());
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libff/algebra/curves/tests/test_groups.cpp:210]: Added a comment about skipping G1 checks

**Review by AntoineRondelet — APPROVED** [review]
Thanks @dtebbs LGTM.

**Pull request merged**